### ISSUE TITLE
When Category or Subcategory NAME is a NUMBER only

### ISSUE THREAD

**Issue by dimalif** [navjottomer/Osclass#38]
If a category or subcategory name is a number only, selecting it, shows zero items.

**There are no results matching ""**

**Comment by webmods-croatia** [comment]
It seems like it searches for category ID instead of the name. Let me verify.

**Comment by webmods-croatia** [comment]
Yep, that's the problem.
I made a category named "Animals". That category ID is 9.
Then I made a category named "9". Clicking on category "9" points to category "Animals" because the script thinks "9" is ID, not a name.

**Comment by webmods-croatia** [comment]
oc-includes/osclass/model/Search.php
```php
        public function addCategory($category = null)
        {
            if( $category == null ) {
                return false;
            }
            if( !is_numeric($category) ) {
                $category  = preg_replace('|/$|','',$category);
                $aCategory = explode('/', $category);
                $category  = Category::newInstance()->findBySlug($aCategory[count($aCategory)-1]);
                if( count($category) == 0 ) {
                    return false;
                }
                $category  = $category['pk_i_id'];
            }
```

If the category isn't numeric, find it by slug. That needs to be changed somehow. First, we need to check if a category with that slug exists.

**Comment by webmods-croatia** [comment]
Or not... "$category" already gives a wrong slug because of permalinks.
Need to check oc-includes/osclass/classes/Rewrite.php

**Comment by webmods-croatia** [comment]
Search page detects category name as page number. The category name is "10", and params have this "[iPage] => 10".

**Comment by webmods-croatia** [comment]
When permalinks are enabled, clicking on category with name "9" points to a category with ID 9.
If there's a subcategory named "10", clicking on it points to a category with ID 9 and page 10.

**Comment by webmods-croatia** [comment]
When permalinks are disabled, everything seems fine.

**Comment by webmods-croatia** [comment]
If the category name is numeric, it automatically finds it by primary key - ID - pk_i_id, instead of the slug.

**Comment by dev-101** [comment]
Is this an issue specific to this dev version or it affected official version, as well? (I assume it did).
Not sure how many users actually need numbers-only categories, though? It may complicate things, not just in rewrite, but I would first look at controller.

If you need aesthetic reasons to show number only, you can change it on the fly with javascript, and keep the title of the category itself with a letter, or something generic e.g. Category 1, Category 2...

**Comment by webmods-croatia** [comment]
Same issue with stock Osclass 3.8.0.
I am checking the search controller currently.

**Comment by webmods-croatia** [comment]
In constructor function of controller/search.php, echoing 
```php
Params::getServerParam('REQUEST_URI', false, false);
```
outputs "/for-sale/animals/10" instead of "9/10".

Seems like the problem is in Rewrite class.

**Comment by dev-101** [comment]
IIRC there's a similar problem with fractions and Custom Fields, but the fix will be complicated, that it is easier left unfixed, and then masked in frontend. Could be the same problem here, haven't had time to look.

**Comment by webmods-croatia** [comment]
> IIRC there's a similar problem with fractions and Custom Fields, but the fix will be complicated, that it is easier left unfixed, and then masked in frontend. Could be the same problem here, haven't had time to look.

I don't actually know about that issue. I will check the Rewrite class anyways.

**Comment by webmods-croatia** [comment]
So apparently, the URL is already modified in init function of Rewrite?

**Comment by dev-101** [comment]
https://github.com/osclass/Osclass/issues/2162

**Comment by webmods-croatia** [comment]
So this will need a lot of work...

> If you need aesthetic reasons to show number only, you can change it on the fly with javascript, and keep the title of the category itself with a letter, or something generic e.g. Category 1, Category 2...

What if we modify category name after the category is added/edited.
If the name is numeric, add "Category " prefix to it making it "Category 2".
Then modify 'osc_category_name' helper to remove "Category " prefix before returning the name.

**Comment by dev-101** [comment]
Well, you can use filters, and add one if there isn't already.

**Comment by webmods-croatia** [comment]
There is an add category hook so we need a name filter only.

**Comment by dev-101** [comment]
Actually, I was thinking the other way around. Store it as 'Category 1', then display it as '1' in theme.

**Comment by webmods-croatia** [comment]
Yes, didn't I say that? ;)
When category saved in admin, store it as "Category 1" and when displayed in theme remove prefix (in osc_category_name()).

**Comment by webmods-croatia** [comment]
This works:

oc-admin/ajax/ajax.php:367
```php
$has_one_title = 1;
if(is_numeric($v)) {
    $v = 'NUMCAT '.$v;
}
$aFieldsDescription[$m[1]][$m[2]] = $v;
$s_text = $v;
```

oc-includes/osclass/helpers/hCategories.php:186
```php
$name = osc_category_field( 's_name' , $locale);
$name = str_replace('NUMCAT ', '', $name);

return $name;
```

**Comment by webmods-croatia** [comment]
Admin form and category model need to be edited as well. Should I proceed with that?

**Comment by dev-101** [comment]
Sorry, didn't understand you, I though you wish to change it before storing in db.

> and when displayed in theme remove prefix (in osc_category_name())

Anyway, that is just one helper, filter is required in category forms...

> Admin form and category model need to be edited as well. Should I proceed with that?

Yes... but I will only add filter and leave all the changes for end users. Not everyone wants/needs this 'NUMCAT' hack (btw. I _like_ the name ;))

**Comment by webmods-croatia** [comment]
> Anyway, that is just one helper, filter is required in category forms...
I will add filters for category names.

> Yes... but I would only add filter and leave all the changes for end users. Not everyone wants/needs this 'NUMCAT' hack (btw. I like the name ;))
What about adding a checkbox in admin?
:)

> But, also, make sure it doesn't break anything later e.g. when editing categories in admin, or search etc. Test it thoroughly.
I will.
